### PR TITLE
Allow users to block and unblock each other

### PR DESCRIPTION
**Pull request by noah-solomon**
## Overview

Context: Our most recent build was rejected from the app store because Pear allows user-generated content, so under app store guidelines, we must allow users to block one another so that they don't have to see filthy chat messages anymore.

This PR allows an authenticated user to block and unblock other users. If a user is blocked, previous matches with this user will not show up, meaning that the blocked user will no longer appear in the messaging view. 

Refer to the SP22 Pear API Specification in Notion for an updated API spec.

## Changes Made

- Add endpoints for blocking and unblocking a user by id (`/api/users/<int:id>/block/`, `/api/users/<int:id>/unblock/`)
- Hide blocked users when seeing "my" matches on `/api/matches/`
- Hide blocked users if they are the current match on `/api/me/` (shows a screen that says "New match coming Sunday")

## Test Coverage

Manually tested in Postman

### CODE REVIEW

**Review by alanna-zhou — APPROVED** [review]
so speedy!! i hate it when appstore is rejecting builds, it's really frustrating. i'll let u go smooth sailing :))

File: src/person/serializers.py [hunk 1]
```diff
@@ -86,6 +92,9 @@ def get_prompts(self, user):
             )
         return prompts
 
+    def get_blocked_users(self, user):
+        return map(lambda u: u.id, user.person.blocked_users.all())
```
Comment (katesliang) [inline src/person/serializers.py:96]: wow lambda functions! very nice!

File: src/person/serializers.py [hunk 2]
```diff
@@ -59,12 +60,17 @@ def get_availability(self, user):
...
         matches = Match.objects.filter(Q(user_1=user) | Q(user_2=user)).order_by(
             "-created_date"
         )
-        if len(matches) == 0:
+        if (
+            len(matches) == 0
+            or matches[0].user_1.id in blocked_ids
+            or matches[0].user_2.id in blocked_ids
+        ):
             return None
```
Comment (katesliang) [inline src/person/serializers.py:72]: will matching algo regenerate matches for ppl who get blocked user D:

File: src/person/views.py [hunk 3]
```diff
@@ -93,3 +93,38 @@ class MassMessageView(generics.GenericAPIView):
...
+    serializer_class = UserSerializer
+    permission_classes = api_settings.CONSUMER_PERMISSIONS
+
+    def post(self, request, id):
+        """Block user by id."""
+        if not User.objects.filter(id=id).exists():
+            return failure_response("User not found.", status.HTTP_404_NOT_FOUND)
+        if (
+            request.user.person.blocked_users.filter(id=id).exists()
+            or request.user.id == id
```
Comment (katesliang) [inline src/person/views.py:108]: should an error alert/message be generated if a user tries to block themselves instead of doing nothing

**Review by katesliang — APPROVED** [review]
very valid!!! hope apple approves now 🥲

File: src/person/views.py (hunk 3 above)
Comment (chalo2000) [reply src/person/views.py:108]: User blocking themself should be an HTTP_403_FORBIDDEN, it doesn't make sense to block yourself and the frontend would not provide a way to do it so this route would never be invoked in that way unless someone is messing with our backend or frontend did something wrong - fix this

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Fix returning a 200 status code for blocking yourself, that should not be allowed everyone should love themselves on our app. Do the same for unblocking.

**Review by chalo2000 — APPROVED** [review]
LGTM :)

**Pull request merged**